controllers/alloydbomni_controller.go

```diff
@@ -0,0 +1,149 @@
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.AlloyDBOmni{}).
+		Owns(&corev1.Secret{}).
+		Complete(r)
+}
+
+func newAlloyDBOmniAdapter(_ *aiven.Client, object client.Object) (serviceAdapter, error) {
+	adbo, ok := object.(*v1alpha1.AlloyDBOmni)
+	if !ok {
+		return nil, fmt.Errorf("object is not of type v1alpha1.AlloyDBOmni")
+	}
+	return &alloyDBOmniAdapter{adbo}, nil
+}
+
+// alloyDBOmniAdapter handles an Aiven AlloyDBOmni service
+type alloyDBOmniAdapter struct {
+	*v1alpha1.AlloyDBOmni
+}
+
+func (a *alloyDBOmniAdapter) getObjectMeta() *metav1.ObjectMeta {
+	return &a.ObjectMeta
+}
+
+func (a *alloyDBOmniAdapter) getServiceStatus() *v1alpha1.ServiceStatus {
+	return &a.Status
+}
+
+func (a *alloyDBOmniAdapter) getServiceCommonSpec() *v1alpha1.ServiceCommonSpec {
+	return &a.Spec.ServiceCommonSpec
+}
+
+func (a *alloyDBOmniAdapter) getUserConfig() any {
+	return a.Spec.UserConfig
+}
+
+func (a *alloyDBOmniAdapter) newSecret(ctx context.Context, s *service.ServiceGetOut) (*corev1.Secret, error) {
+	stringData := map[string]string{
+		"HOST":         s.ServiceUriParams["host"],
+		"PORT":         s.ServiceUriParams["port"],
+		"DATABASE":     s.ServiceUriParams["dbname"],
+		"USER":         s.ServiceUriParams["user"],
+		"PASSWORD":     s.ServiceUriParams["password"],
+		"SSLMODE":      s.ServiceUriParams["sslmode"],
+		"DATABASE_URI": s.ServiceUri,
+	}
+
+	return newSecret(a, stringData, true), nil
+}
+
+func (a *alloyDBOmniAdapter) getServiceType() string {
+	return "alloydbomni"
+}
+
+func (a *alloyDBOmniAdapter) getDiskSpace() string {
+	return a.Spec.DiskSpace
+}
+
+func (a *alloyDBOmniAdapter) performUpgradeTaskIfNeeded(ctx context.Context, avn avngen.Client, old *service.ServiceGetOut) error {
+	return nil
+}
+
+func (a *alloyDBOmniAdapter) createOrUpdateServiceSpecific(ctx context.Context, avnGen avngen.Client, old *service.ServiceGetOut) error {
+	// Get current credentials state
+	currentCreds, err := avnGen.AlloyDbOmniGoogleCloudPrivateKeyIdentify(ctx, a.Spec.Project, a.Name)
+	if err != nil {
+		return fmt.Errorf("failed to get current credentials: %w", err)
+	}
+
+	// Handle removal case first
+	if a.Spec.ServiceAccountCredentials == "" {
+		if currentCreds.PrivateKeyId != "" {
+			if _, err := avnGen.AlloyDbOmniGoogleCloudPrivateKeyRemove(ctx, a.Spec.Project, a.Name); err != nil {
+				return fmt.Errorf("failed to remove credentials: %w", err)
+			}
+		}
+		return nil
+	}
+
+	// Validate new credentials
+	if err := validateCredentials(a.Spec.ServiceAccountCredentials); err != nil {
+		return fmt.Errorf("invalid credentials: %w", err)
+	}
+
+	// Parse credentials to get key ID
+	var credsMap map[string]interface{}
+	if err := json.Unmarshal([]byte(a.Spec.ServiceAccountCredentials), &credsMap); err != nil {
+		return fmt.Errorf("failed to parse credentials: %w", err)
+	}
+
+	// Update only if key ID changed
+	if newKeyID := credsMap["private_key_id"].(string); newKeyID != currentCreds.PrivateKeyId {
```

Comment:
So if `ServiceAccountCredentials` is an empty string, this might panic.

```go
type creds struct {
    PrivateKeyID string `json:"private_key_id"`
}

c := new(creds)
err := json.Unmarshal([]byte(a.Spec.ServiceAccountCredentials), c)

if c.PrivateKeyId != currentCreds.PrivateKeyId {
    ...
}
```

Reply:
Resolved in latest changeset